src/Processors/Formats/Impl/ArrowBlockInputFormat.cpp

```diff
@@ -0,0 +1,115 @@
+#include "ArrowBlockInputFormat.h"
+#if USE_ARROW
+
+#include <Formats/FormatFactory.h>
+#include <IO/ReadBufferFromMemory.h>
+#include <IO/WriteBufferFromString.h>
+#include <IO/WriteHelpers.h>
+#include <IO/copyData.h>
+#include <arrow/api.h>
+#include <arrow/io/memory.h>
+#include <arrow/ipc/reader.h>
+#include <arrow/status.h>
+#include "ArrowColumnToCHColumn.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_READ_ALL_DATA;
+    extern const int BAD_ARGUMENTS;
+}
+
+
+ArrowBlockInputFormat::ArrowBlockInputFormat(ReadBuffer & in_, Block header_) : IInputFormat(std::move(header_), in_)
+{
+}
+
+Chunk ArrowBlockInputFormat::generate()
+{
+    Chunk res;
+
+    const auto & header = getPort().getHeader();
+
+    if (!in.eof())
+    {
+        if (row_group_current < row_group_total)
+            throw Exception{"Got new data, but data from previous chunks was not read " +
+                            std::to_string(row_group_current) + "/" + std::to_string(row_group_total),
+                            ErrorCodes::CANNOT_READ_ALL_DATA};
+
+        file_data.clear();
+        {
+            WriteBufferFromString file_buffer(file_data);
+            copyData(in, file_buffer);
+        }
+
+        std::unique_ptr<arrow::Buffer> local_buffer = std::make_unique<arrow::Buffer>(file_data);
```

Comment:
Is it possible to read data without excessive copying and allocating O(n) memory?
That's extremely suboptimal.